Chainlink OCR aggregators as reporters 🚀

alexroan opened the pull request.

alexroan committed:
Chainlink OCR aggregators as reporters

alexroan committed:
UniswapConfig test getTokenConfigByReporter

alexroan committed:
Reduced contract size by reducing number of configs allowed

alexroan commented on contracts/Uniswap/UniswapConfig.sol:
```diff
@@ -25,12 +25,13 @@ contract UniswapConfig {
     /// @notice The max number of tokens this contract is hardcoded to support
     /// @dev Do not change this variable without updating all the fields throughout the contract.
-    uint public constant maxTokens = 30;
+    uint public constant maxTokens = 20;
```
Reduced the max number of tokens to reduce the size of the resulting contract to fit within the limit.

There are currently only 13 being used in production. [source here](https://etherscan.io/address/0x922018674c12a7f0d394ebeef9b58f186cde13c1#readContract)

alexroan committed:
NonReporterPricesTests for OCR Aggregator prices

alexroan committed:
Config testing new max config limit

alexroan committed:
UniswapAnchoredView.validate tests

alexroan committed:
UniswapAnchoredView.getUnderlyingPrice tests for Chainlink reporters

alexroan committed:
UniswapAnchoredViewTest fixes

alexroan committed:
PostRealWorldPricesTest init, basic scenario and PriceUpdated tests

alexroan committed:
Fixed last PortRealWorldPricesTest

alexroan committed:
Rearrange functions

alexroan committed:
28 supported markets in config

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -140,52 +130,42 @@ contract UniswapAnchoredView is UniswapConfig {
-            emit PriceUpdated(symbol, reporterPrice);
+            prices[config.symbolHash] = reporterPrice;
+            emit PriceUpdated(config.symbolHash, reporterPrice);
+            valid = true;
```
We don't use this return value, though it also doesn't hurt to have it here.

alexroan replied:
I'm happy to remove, but it makes sense from a contextual perspective.

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -211,14 +191,14 @@ contract UniswapAnchoredView is UniswapConfig {
      * @param conversionFactor 1e18 if seeking the ETH price, and a 6 decimal ETH-USDC price in the case of other assets
      */
-    function fetchAnchorPrice(string memory symbol, TokenConfig memory config, uint conversionFactor) internal virtual returns (uint) {
+    function fetchAnchorPrice(bytes32 symbolHash, TokenConfig memory config, uint conversionFactor) internal virtual returns (uint) {
```
why this modification?

alexroan replied:
We never use the string version. `UniswapConfig` stores as `immutable bytes32 symbolHash`'s.

The string version was used for readability (I think?) in events. Every function that did accept a symbol string immediately converted it into a hash to use it, but we've removed those functions now.

kaleofduty replied:
Thanks for the context. Sgtm

kaleofduty commented on contracts/Uniswap/UniswapConfig.sol:
```diff
@@ -711,6 +764,20 @@ contract UniswapConfig {
+     * @param reporter The address of the reporter of the config to get
+     * @return The config object
+     */
+    function getTokenConfigByReporter(address reporter) public view returns (TokenConfig memory) {
```
I'm wondering whether it would be clearer to use `aggregator` instead of `reporter`. Thoughts?

alexroan replied:
`reporter` was kept for the sake of using terms and concepts familiar to the Compound community. I think I'd rather keep as is, since it would change the naming conventions in a few other places like `invalidateReporter()`.

kaleofduty replied:
Makes sense 👍

alexroan committed:
Chainlink price feed failovers

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -7,12 +7,18 @@ import "./UniswapConfig.sol";
     uint acc;
 }
 
+struct OfficialPriceData {
```
This uses two words. If we can use a uint248, it will into one word.

kaleofduty replied:
Why "OfficialPriceData" instead of "PriceData"

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -254,12 +259,23 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-        reporterInvalidated[reporterAddr] = true;
-        emit ReporterInvalidated(reporterAddr);
+    function deactivateFailover(bytes32 symbolHash) external onlyOwner() {
+        require(prices[symbolHash].failoverActive == true, "Already deactivated");
```
```suggestion
        require(prices[symbolHash].failoverActive, "Already deactivated");
```

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -254,12 +259,23 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
+     * @dev Only the owner can call this function
+     */
+    function activateFailover(bytes32 symbolHash) external onlyOwner() {
+        require(prices[symbolHash].failoverActive == false, "Already activated");
```
```suggestion
        require(!prices[symbolHash].failoverActive, "Already activated");
```

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -153,11 +161,8 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-            emit PriceUpdated(config.symbolHash, anchorPrice);
-        } else if (isWithinAnchor(reporterPrice, anchorPrice)) {
-            prices[config.symbolHash] = reporterPrice;
+        if (isWithinAnchor(reporterPrice, anchorPrice)) {
```
This is probably fine, but I want to point out that this will mean that we `emit PriceUpdated` even while the failover is active.

alexroan replied:
I think that is OK.

If failover was activated due to stale aggregator, these events firing again would signal that it is safe to `deactivateFailover` and roll back to the anchored prices.

alexroan committed:
getUnderlyingPrice handles failovers with different decimals

alexroan committed:
Storage saving and linting

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -134,58 +140,44 @@ contract UniswapAnchoredView is UniswapConfig {
-            prices[symbolHash] = reporterPrice;
-            emit PriceUpdated(symbol, reporterPrice);
+        if (isWithinAnchor(reporterPrice, anchorPrice)) {
+            prices[config.symbolHash].price = uint248(reporterPrice);
```
Safe cast?

alexroan committed:
Safe cast price

alexroan committed:
Removed CL Price Feed failover mechanism

alexroan committed:
Activate and Deactivate failover to uniswap. Also, storage saving on prices

alexroan committed:
Convert reported price to 6 decimals before storing

alexroan committed:
pokeFailedOverPrice, using 0.6.12 and refactor

alexroan committed:
Test fixes

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -140,52 +135,81 @@ contract UniswapAnchoredView is UniswapConfig {
+
+        PriceData memory priceData = prices[config.symbolHash];
+        if (priceData.failoverActive) {
+            require(anchorPrice <= 2**248, "Anchor price too large");
```
```suggestion
            require(anchorPrice < 2**248, "Anchor price too large");
```

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -140,52 +135,81 @@ contract UniswapAnchoredView is UniswapConfig {
+            prices[config.symbolHash].price = uint248(anchorPrice);
+            emit PriceUpdated(config.symbolHash, anchorPrice);
+        } else if (isWithinAnchor(reportedPrice, anchorPrice)) {
+            require(reportedPrice <= 2**248, "Reported price too large");
```
```suggestion
            require(reportedPrice < 2**248, "Reported price too large");
```

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -140,52 +135,81 @@ contract UniswapAnchoredView is UniswapConfig {
+        require(priceData.failoverActive, "Failover must be active");
+        TokenConfig memory config = getTokenConfigBySymbolHash(symbolHash);
+        uint anchorPrice = calculateAnchorPriceFromEthPrice(config);
+        require(anchorPrice <= 2**248, "Anchor price too large");
```
```suggestion
        require(anchorPrice < 2**248, "Anchor price too large");
```

alexroan committed:
Stricter safe casting

alexroan committed:
Non-zero price value on construction

alexroan committed:
Require reporter address in constructor

Explicit testing for setting UniswapConfig variables corectly

alexroan committed:
Notes for clarity on constructor values and access control

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -1,32 +1,32 @@
-    /// @notice The Open Oracle Price Data contract
-    OpenOraclePriceData public immutable priceData;
+contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Ownable {
+    using FixedPoint for *;
```
Any way we can restrict the type here?

alexroan replied:
We can, but this is an unchanged line that git seems to have deemed changed (see removed line 16 above). `using FixedPoint for *;` is used in a few contracts throughout the repo.

kaleofduty replied:
Ah, I see. Good to leave as is in that case, thank you.

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -93,7 +90,9 @@ contract UniswapAnchoredView is UniswapConfig {
             address uniswapMarket = config.uniswapMarket;
             if (config.priceSource == PriceSource.REPORTER) {
                 require(uniswapMarket != address(0), "reported prices must have an anchor");
+                require(config.reporter != address(0), "reported price much have a reporter");
```
typo

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -140,52 +139,82 @@ contract UniswapAnchoredView is UniswapConfig {
-     * @param signatures The signatures for the corresponding messages
-     * @param symbols The symbols to compare to anchor for authoritative reading
+     * @notice This is called by the reporter whenever a new price is posted on-chain
+     * @dev called by AccessControlledOffChainAggregator
```
```suggestion
     * @dev called by AccessControlledOffchainAggregator
```

alexroan committed:
Update contracts/Uniswap/UniswapAnchoredView.sol

Co-authored-by: kaleofduty <59616916+kaleofduty@users.noreply.github.com>

alexroan committed:
Fixed typo

alexroan committed:
Clarify underlying price math comment

alexroan committed:
Poke failoved over price when failover activated

Scaling factors clearly documented for FIXED_ETH and FIXED_USD. fetchEthPrice rename to fetchEthAnchorPrice

kaleofduty commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -133,8 +133,10 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
+         // Comptroller needs prices in the format: ${raw price} * 1e36 / baseUnit
+         // The baseUnit of an asset is the amount of the smallest denomination of that asset per whole.
+         // For example, the baseUnit of ETH is 1e18.
+         // Since the prices in this view have 6 decimals, we must scale them by 1e36 - 1e6 - baseUnit
```
```suggestion
         // Since the prices in this view have 6 decimals, we must scale them by 1e(36 - 6)/baseUnit
```

alexroan committed:
Math comment clarification.

Co-authored-by: kaleofduty <59616916+kaleofduty@users.noreply.github.com>

coburncoburn commented:
since this is going live in a few hours, time to merge the PR?
